### PR TITLE
move Trunc, Cot, Round, iszero functions to datafusion-functions

### LINKED ISSUE #9859: move Trunc, Cot, Round, iszero functions to datafusion-functions crate
### Is your feature request related to a problem or challenge?

As part of https://github.com/apache/arrow-datafusion/issues/9285 move the Trunc, Cot, Round, iszero functions to the new datafusion-functions/math crate in the new structure.

### Describe the solution you'd like

Functions are migrated to the new crate and use the new patterns as described in https://github.com/apache/arrow-datafusion/issues/9286 & https://github.com/apache/arrow-datafusion/pull/9216

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #9859

## Rationale for this change

Function migration for Trunc, cot, round and iszero functions as part of https://github.com/apache/arrow-datafusion/issues/9285, 

## What changes are included in this PR?

Code, tests. math/mod.rs had to be redone to account for trunc taking a Vec<Expr> argument.

## Are these changes tested?

Yes.

## Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion/functions/src/math/round.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        return exec_err!(
+            "round function requires one or two arguments, got {}",
+            args.len()
+        );
+    }
+
+    let mut decimal_places = ColumnarValue::Scalar(ScalarValue::Int64(Some(0)));
+
+    if args.len() == 2 {
+        decimal_places = ColumnarValue::Array(args[1].clone());
```
Comment (Omega359) [inline datafusion/functions/src/math/round.rs:100]: The whole ColumnarArray -> ArrayRef -> ColumnarArray in this code really should be cleaned up in another PR imho.

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @Omega359 .

There was some pretty gnarly merge conflicts to handle due to  https://github.com/apache/arrow-datafusion/pull/9983

I did this locally as part of my review and made a PR to your branch https://github.com/Omega359/arrow-datafusion/pull/2 in case you would ilke to use that

Thanks again

**Comment by Omega359** [comment]
> There was some pretty gnarly merge conflicts to handle due to #9983
> 
> I did this locally as part of my review and made a PR to your branch [Omega359#2](https://github.com/Omega359/arrow-datafusion/pull/2) in case you would ilke to use that

Yeah, I was expecting that depending on which landed first. I'll check out your merge tonight - thanks!

**Comment by alamb** [comment]
Check config failure seems unrelated: https://github.com/apache/arrow-datafusion/actions/runs/8606494312/job/23585168125?pr=10000

> /usr/bin/docker exec  6cacef2be1f3e8731bdd8120b9ceb79bfc495030e312805a9aedd199a82efc2b sh -c "cat /etc/*release | grep ^ID"
> /__e/node20/bin/node: error while loading shared libraries: /lib/x86_64-linux-gnu/libstdc++.so.6: invalid ELF header


I'll try and rerun it when the rest of the tests pass

**Comment by Omega359** [comment]
> I'll try and rerun it when the rest of the tests pass
Thanks, all the tests I have are passing locally.

**Comment by alamb** [comment]
Reopened restart checks

**Comment by alamb** [comment]
Just let me know when you are happy with this PR @Omega359  and I'll merge it in

**Comment by Omega359** [comment]
I believe this PR is complete and ready to merge @alamb, thx for the review and handling the merge conflict.

**Comment by alamb** [comment]
Thanks again @Omega359
